### ISSUE TITLE
inconsistent grads between two types of allgather

### ISSUE THREAD

**Issue by gameofdimension** [pytorch/pytorch#153016]
### 🐛 Describe the bug

I've observed a gradient discrepancy between two PyTorch all-gather implementations: one using the DTensor API, and the other using all_gather_tensor_autograd. My goal is to implement a correct autograd-compatible all-gather operation, but I'm unsure which implementation (if either) produces the right gradients.

```python
import os
from datetime import timedelta

import torch
from torch.distributed import init_device_mesh
from torch.distributed.tensor import DTensor, Replicate, Shard


def make_data(gen, dtype):
    data = torch.rand(4, 5, requires_grad=True, generator=gen).to(
        device=torch.cuda.current_device(),
        dtype=dtype,
    )
    data.retain_grad()
    return data


def init_distributed():
    local_rank = int(os.environ["LOCAL_RANK"])
    torch.cuda.set_device(local_rank)

    init_timeout_seconds = 300
    torch.distributed.init_process_group(
        backend="nccl",
        timeout=timedelta(seconds=init_timeout_seconds),
    )


def allgather1(mesh, data: torch.Tensor):
    return DTensor.from_local(
        data, mesh, [Shard(0)]
    ).redistribute(
        mesh, [Replicate()]
    ).to_local()


def allgather2(mesh, data):
    return torch.distributed._functional_collectives.all_gather_tensor_autograd(
        data, gather_dim=0, group=mesh.get_group())


def square_fb(out):
    loss = (out**2).sum()
    loss.backward()
    torch.cuda.synchronize()
    return loss


def check1(mesh, data):
    out = allgather1(mesh, data)
    loss = square_fb(out)
    # 1. grad is 2*input
    assert torch.allclose(data.grad, 2 * data)
    return out, loss


def check2(mesh, data):
    out = allgather2(mesh, data)
    loss = square_fb(out)
    # 2. grad is 2*world_size*input
    assert torch.allclose(data.grad, 2 * torch.distributed.get_world_size() * data)
    return out, loss


def check_allgather():
    init_distributed()
    world_size = torch.distributed.get_world_size()
    mesh = init_device_mesh(device_type="cuda", mesh_shape=[world_size])

    dtype = torch.float32
    rank = torch.distributed.get_rank()

    data1 = make_data(torch.Generator().manual_seed(10 + rank), dtype)
    data2 = make_data(torch.Generator().manual_seed(10 + rank), dtype)
    assert torch.allclose(data1, data2)

    ag1, loss1 = check1(mesh, data1)
    ag2, loss2 = check2(mesh, data2)
    assert torch.allclose(ag1, ag2)  # two allgathers produce the same result
    assert torch.allclose(loss1, loss2)  # both forwards are equal too

    print("all check passed")


if __name__ == "__main__":
    check_allgather()
```

**to run the script**
```
torchrun --nproc-per-node=8 repro.py

### Versions

Collecting environment information...
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.11.9 (main, May 14 2024, 09:36:59) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.0-136.36.0.112.4.oe2203sp1.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A800-SXM4-80GB
GPU 1: NVIDIA A800-SXM4-80GB
GPU 2: NVIDIA A800-SXM4-80GB
GPU 3: NVIDIA A800-SXM4-80GB
GPU 4: NVIDIA A800-SXM4-80GB
GPU 5: NVIDIA A800-SXM4-80GB
GPU 6: NVIDIA A800-SXM4-80GB
GPU 7: NVIDIA A800-SXM4-80GB

Nvidia driver version: 535.161.07
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8352Y CPU @ 2.20GHz
Stepping:                        6
CPU MHz:                         2799.998
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4400.00
Virtualization:                  VT-x
L1d cache:                       3 MiB
L1i cache:                       2 MiB
L2 cache:                        80 MiB
L3 cache:                        96 MiB
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Vulnerable
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.0
[pip3] torchdiffeq==0.2.5
[pip3] torchsde==0.2.6
[pip3] torchvision==0.22.0
[pip3] triton==3.3.0
[conda] Could not collect

**Comment by gameofdimension** [comment]
@H-Huang @awgu @wanchaol @fegin 

Would you please review this issue? I’d appreciate your insights on whether there’s a bug here or if I’m misunderstanding something.

**Comment by gameofdimension** [comment]
@wconstab would you mind taking a look

**Comment by gameofdimension** [comment]
The DTensor-based `allgather1` was absolutely incorrect; this can be easily verified by replacing the loss function with the one below.

```
def dummy_loss(out):
    rank = torch.distributed.get_rank()
    loss = (out*(1+rank)).sum()
    loss.backward()
    torch.cuda.synchronize()
    return loss
```
